File: gamspy/simple_steel_data.py
import numpy as np
import pandas as pd
import h5py
import os
from gamspy import Container, Set, Parameter, Sum

def generate_data():
    print("Generating data using GAMSPy...")
    
    # Initialize Container
    m = Container()

    # ------------------------------------------------------------------------------
    # 1. DECLARE PARAMETERS, SETS, DATA
    # ------------------------------------------------------------------------------

    # Sets
    i_recs = [f"i{k}" for k in range(1, 16)]
    i = Set(m, "i", records=i_recs, description="Steel plants")
    
    f_recs = ["iore", "coal", "scrp", "elec", "ngas"]
    f = Set(m, "f", records=f_recs, description="Factors")
    
    t_recs = ["bof", "eaf", "dri"]
    t = Set(m, "t", records=t_recs, description="Technology type")

    # Mapping plants to technologies
    it_recs = []
    for k in range(1, 16):
        plant = f"i{k}"
        if k <= 5:
            tech = "bof"
        elif k <= 10:
            tech = "eaf"
        else:
            tech = "dri"
        it_recs.append((plant, tech))

    it = Set(
        m,
        "it",
        domain=[i, t],
        records=it_recs,
        description="Link between plants and technologies",
    )

    # Parameters
    # Load from CSVs in data/raw
    raw_data_dir = "data/raw"

    # abar: Factor use per ton
    # CSV has columns: t, f, value
    abar_df = pd.read_csv(os.path.join(raw_data_dir, "abar.csv"))
    abar = Parameter(
        m, "abar", domain=[t, f], records=abar_df,
        description="Factor use per ton crude steel production"
    )

    # vbar: Reference factor market prices
    # CSV has columns: f, value
    vbar_df = pd.read_csv(os.path.join(raw_data_dir, "vbar.csv"))
    vbar = Parameter(
        m, "vbar", domain=[f], records=vbar_df,
        description="Reference factor market prices"
    )

    # rho: Price elasticity in factor supply
    # CSV has columns: f, value
    rho_df = pd.read_csv(os.path.join(raw_data_dir, "rho.csv"))
    rho = Parameter(
        m, "rho", domain=[f], records=rho_df,
        description="Price elasticity in factor supply"
    )

    # kappa: Carbon emissions coefficient
    # CSV has columns: f, value
    kappa_df = pd.read_csv(os.path.join(raw_data_dir, "kappa.csv"))
    kappa = Parameter(
        m, "kappa", domain=[f], records=kappa_df,
        description="Carbon emissions coefficient"
    )

    epsilon = Parameter(
        m, "epsilon", records=-0.3, description="Price elasticity of steel demand"
    )

    # Random Generation Data
    np.random.seed(42)
    n_plants = 15
    n_factors = 5

    # Flatten the factor records to a list of strings for iteration
    # f.records is a DataFrame
    f_list = f_recs

    tau_data = np.random.uniform(0, 0.1, size=(n_plants, n_factors))
    tau_df = pd.DataFrame(
        [
            (f"i{r + 1}", f_name, tau_data[r, c])
            for r in range(n_plants)
            for c, f_name in enumerate(f_list)
        ],
        columns=["i", "f", "val"],
    )
    tau = Parameter(
        m, "tau", domain=[i, f], records=tau_df,
        description="Plant-specific additional cost (e.g. transport)"
    )

    ylim_data = np.random.uniform(10, 15, size=n_plants)
    ylim_df = pd.DataFrame(
        [(f"i{r + 1}", ylim_data[r]) for r in range(n_plants)], columns=["i", "val"]
    )
    ylim = Parameter(
        m, "ylim", domain=[i], records=ylim_df,
        description="Plant-specific production capacity"
    )

    ybar_init_data = ylim_data - np.random.uniform(0, 5, size=n_plants)
    ybar_df = pd.DataFrame(
        [(f"i{r + 1}", ybar_init_data[r]) for r in range(n_plants)],
        columns=["i", "val"],
    )
    ybar = Parameter(
        m, "ybar", domain=[i], records=ybar_df,
        description="Reference output by plant"
    )

    # dbar = sum(i, ybar(i))
    # We can calculate this in python or via gamspy
    dbar_val = ybar_df["val"].sum()
    dbar = Parameter(m, "dbar", records=dbar_val, description="Reference aggregate demand")

    print(f"Data generated. dbar: {dbar_val}")
    
    # Define Policy parameter (needed for model structure)
    chi = Parameter(m, "chi", records=0, description="Emissions reduction target")

    # ------------------------------------------------------------------------------
    # EXPORT TO GDX
    # ------------------------------------------------------------------------------
    if not os.path.exists("data"):
        os.makedirs("data")
    
    gdx_path = os.path.abspath("data/generated/steel_data.gdx")
    m.write(gdx_path)
    print(f"GDX file written to: {gdx_path}")

    # ------------------------------------------------------------------------------
    # EXPORT TO HDF5 (Structure matching R export)
    # ------------------------------------------------------------------------------
    if not os.path.exists("data/generated"):
        os.makedirs("data/generated")
        
    h5_path = "data/generated/steel_data.h5"
    
    with h5py.File(h5_path, 'w') as f:
        # Sets
        f.create_dataset("sets/i", data=np.array(i_recs, dtype='S'))
        f.create_dataset("sets/f", data=np.array(f_recs, dtype='S'))
        f.create_dataset("sets/t", data=np.array(t_recs, dtype='S'))
        
        # it (DataFrame)
        # Convert list of tuples to dataframe for easier saving or numpy structured array
        it_df = pd.DataFrame(it_recs, columns=["i", "t"])
        # Save as compound dataset or strings. 
        # R's hdf5r often saves dataframes as compound types. 
        # Simplest for cross-compat: save columns as separate arrays or struct array.
        # Let's use numpy struct array which h5py supports well
        it_arr = np.array([tuple(x) for x in it_recs], dtype=[('i', 'S10'), ('t', 'S10')])
        f.create_dataset("sets/it", data=it_arr)

        # Parameters
        # Helpers
        def save_param(path, df, dtype):
            # Convert dataframe to numpy structured array
            recs = df.to_records(index=False)
            # Ensure string cols are S type
            new_dtype = []
            for name, d in dtype:
                new_dtype.append((name, d))
            
            # Cast
            arr = np.array(recs, dtype=new_dtype)
            f.create_dataset(path, data=arr)

        save_param("params/abar", abar_df, [('t', 'S10'), ('f', 'S10'), ('value', 'f8')])
        save_param("params/vbar", vbar_df, [('f', 'S10'), ('value', 'f8')])
        save_param("params/rho", rho_df, [('f', 'S10'), ('value', 'f8')])
        save_param("params/kappa", kappa_df, [('f', 'S10'), ('value', 'f8')])
        
        save_param("params/tau", tau_df.rename(columns={"val":"value"}), [('i', 'S10'), ('f', 'S10'), ('value', 'f8')])
        save_param("params/ylim", ylim_df.rename(columns={"val":"value"}), [('i', 'S10'), ('value', 'f8')])
        save_param("params/ybar", ybar_df.rename(columns={"val":"value"}), [('i', 'S10'), ('value', 'f8')])
        
        # Scalars
        f.create_dataset("scalars/dbar", data=dbar_val)
        f.create_dataset("scalars/epsilon", data=-0.3)
        f.create_dataset("scalars/chi", data=0.0)

    print(f"HDF5 file written to: {h5_path}")

if __name__ == "__main__":
    generate_data()

File: R/simple_steel_data.R
# R/simple_steel_data.R

library(data.table)
library(gamstransfer)

set.seed(42) # For reproducibility


# Sets -------------------------------------------------------------------

i <- paste0("i", 1:15)

f_names <- c(
  "iore" = "Iron ore (tonnes)",
  "coal" = "Coking coal (tonnes)",
  "scrp" = "Steel scrap (tonnes)",
  "elec" = "Electricity (MWh)",
  "ngas" = "Natural gas (tonnes)"
)

f <- names(f_names)

t_names <- c(
  "bof" = "BF-BOF Route",
  "eaf" = "Scrap-EAF Route",
  "dri" = "DRI-EAF Route"
)

t_tech <- names(t_names)

# Link between plants and technologies
# (i1 * i5).bof, (i6 * i10).eaf, (i11 * i15).dri
it <- data.table(
  i = i,
  t = c(rep("bof", 5), rep("eaf", 5), rep("dri", 5))
)

# Parameters -------------------------------------------------------------

# abar: Factor use per ton crude steel production
abar <- fread("data/raw/abar.csv")

# vbar: Reference factor market prices
vbar <- fread("data/raw/vbar.csv")

# rho: Price elasticity in factor supply
rho <- fread("data/raw/rho.csv")

# kappa: Carbon emissions coefficient
kappa <- fread("data/raw/kappa.csv")

epsilon <- -0.3


# Random Data Generation -------------------------------------------------

# tau: Plant-specific additional cost (e.g. transport)
# tau(i, f) = uniform(0, 0.1)
tau <- CJ(i = i, f = f)
tau[, value := runif(.N, min = 0, max = 0.1)]

# ylim: Plant-specific production capacity
# ylim(i) = uniform(10, 15)
ylim <- data.table(i = i)
ylim[, value := runif(.N, min = 10, max = 15)]

# ybar: Reference output by plant
# ybar(i) = ylim(i) - uniform(0, 5)
ybar <- copy(ylim)
ybar[, value := value - runif(.N, min = 0, max = 5)]


# dbar: Reference aggregate demand
# dbar = sum(i, ybar(i))
dbar <- sum(ybar$value)

# Policy Parameter
chi <- 0

# Output to check
print("Data generation complete.")
print(paste("dbar:", dbar))

# ------------------------------------------------------------------------------
# HDF5 EXPORT (Structured)
# ------------------------------------------------------------------------------

if (!require("hdf5r")) install.packages("hdf5r", repos = "http://cran.us.r-project.org")
library(hdf5r)

# --- HDF5 Test: Can we create any HDF5 file? ---
tryCatch({
  test_file <- tempfile(fileext = ".h5")
  test_h5 <- H5File$new(test_file, mode = "w")
  test_h5[["test_data"]] <- 1:10
  test_h5$close_all()
  message("HDF5 test file created successfully at: ", test_file)
  file.remove(test_file)
  message("HDF5 test file removed.")
}, error = function(e) {
  stop("Failed to create basic HDF5 test file. HDF5 library or hdf5r package might be misconfigured: ", e$message)
})
# --- End HDF5 Test ---

# Create directory if not exists
if (!dir.exists("data/generated")) dir.create("data/generated", recursive = TRUE)

h5_file <- "data/generated/steel_data.h5"
if (file.exists(h5_file)) file.remove(h5_file)

file.h5 <- H5File$new(h5_file, mode = "w")

# --- 1. Sets ---
# Explicitly create groups
sets_group <- file.h5$create_group("sets")

# Write simple sets as 1D arrays
sets_group$create_dataset("i", i)
sets_group$create_dataset("f", f)
sets_group$create_dataset("t", t_tech)

# Write mapping 'it' as a dataframe/compound type
sets_group$create_dataset("it", it)

# --- 2. Parameters ---
params_group <- file.h5$create_group("params")

# We write them as they are (DataTables).
params_group$create_dataset("abar", abar)
params_group$create_dataset("vbar", vbar)
params_group$create_dataset("rho", rho)
params_group$create_dataset("kappa", kappa)

params_group$create_dataset("tau", tau)
params_group$create_dataset("ylim", ylim)
params_group$create_dataset("ybar", ybar)

# --- 3. Scalars ---
scalars_group <- file.h5$create_group("scalars")

scalars_group$create_dataset("dbar", dbar)
scalars_group$create_dataset("epsilon", epsilon)
scalars_group$create_dataset("chi", chi)
file.h5$close_all()

print(paste("All data written to single HDF5 file:", h5_file))

# ------------------------------------------------------------------------------
# GDX EXPORT
# ------------------------------------------------------------------------------

m <- Container$new()

# Sets
# For sets with descriptions, we format records as data frame 
# with 'uni' and 'element_text'
i_set <- Set$new(
  m,
  "i",
  records = i,
  description = "Steel plants"
)

f_records <- data.frame(
  uni = names(f_names),
  element_text = as.character(f_names)
)

f_set <- Set$new(
  m,
  "f",
  records = f_records,
  description = "Factors"
)

t_records <- data.frame(
  uni = names(t_names),
  element_text = as.character(t_names)
)

t_set <- Set$new(
  m,
  "t",
  records = t_records,
  description = "Technology type"
)

it_set <- Set$new(
  m,
  "it",
  domain = c(i_set, t_set),
  records = it,
  description = "Link between plants and technologies"
)

# Parameters
abar_p <- Parameter$new(
  m,
  "abar",
  domain = c(t_set, f_set),
  records = abar,
  description = "Factor use per ton crude steel production"
)

vbar_p <- Parameter$new(
  m,
  "vbar",
  domain = f_set,
  records = vbar,
  description = "Reference factor market prices"
)

rho_p <- Parameter$new(
  m,
  "rho",
  domain = f_set,
  records = rho,
  description = "Price elasticity in factor supply"
)

kappa_p <- Parameter$new(
  m,
  "kappa",
  domain = f_set,
  records = kappa,
  description = "Carbon emissions coefficient"
)

epsilon_p <- Parameter$new(
  m,
  "epsilon",
  records = epsilon,
  description = "Price elasticity of steel demand"
)

tau_p <- Parameter$new(
  m,
  "tau",
  domain = c(i_set, f_set),
  records = tau,
  description = "Plant-specific additional cost (e.g. transport)"
)

ylim_p <- Parameter$new(
  m,
  "ylim",
  domain = i_set,
  records = ylim,
  description = "Plant-specific production capacity"
)

ybar_p <- Parameter$new(
  m,
  "ybar",
  domain = i_set,
  records = ybar,
  description = "Reference output by plant"
)

dbar_p <- Parameter$new(
  m,
  "dbar",
  records = dbar,
  description = "Reference aggregate demand"
)

chi_p <- Parameter$new(
  m,
  "chi",
  records = chi,
  description = "Emissions reduction target"
)

# Write GDX
if (!dir.exists("data")) dir.create("data", recursive = TRUE)
gdx_path <- "data/generated/steel_data.gdx"
m$write(gdx_path)
print(paste("GDX file written to:", gdx_path))